src/matchers/numbers.lua

```diff
@@ -1,22 +1,128 @@
+local prettyValue = require("src.utils.prettyValue")
+local assertType = require("src.asserts.type")
+
+local INFINITY = math.huge
+
 ---@type lest.Matcher
-local function toBeCloseTo(ctx, received, expected, numDigits) end
+local function toBeCloseTo(ctx, received, expected, numDigits)
+	assertType(received, "number")
+	assertType(expected, "number")
+	assertType(numDigits, "number")
+
+	local expectedDiff = math.pow(10, -numDigits) / 2
+	local receivedDiff = math.abs(expected - received)
+	return {
+		pass = receivedDiff < expectedDiff,
+		message = string.format(
+			"Expected %s to%sbe close to %s within %s digits",
+			prettyValue(received),
+			ctx.inverted and " not " or " ",
+			prettyValue(expected),
+			prettyValue(numDigits)
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeGreaterThan(ctx, received, expected) end
+local function toBeGreaterThan(ctx, received, expected)
+	assertType(received, "number")
+	assertType(expected, "number")
+
+	return {
+		pass = received > expected,
+		message = string.format(
+			"Expected %s to%sbe greater than %s",
+			prettyValue(received),
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeGreaterThanOrEqual(ctx, received, expected) end
+local function toBeGreaterThanOrEqual(ctx, received, expected)
+	assertType(received, "number")
+	assertType(expected, "number")
+
+	return {
+		pass = received >= expected,
+		message = string.format(
+			"Expected %s to%sbe greater than or equal to %s",
+			prettyValue(received),
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeLessThan(ctx, received, expected) end
+local function toBeLessThan(ctx, received, expected)
+	assertType(received, "number")
+	assertType(expected, "number")
+
+	return {
+		pass = received < expected,
+		message = string.format(
+			"Expected %s to%sbe less than %s",
+			prettyValue(received),
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeLessThanOrEqual(ctx, received, expected) end
+local function toBeLessThanOrEqual(ctx, received, expected)
+	assertType(received, "number")
+	assertType(expected, "number")
+
+	return {
+		pass = received <= expected,
+		message = string.format(
+			"Expected %s to%sbe less than or equal to %s",
+			prettyValue(received),
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeNaN(ctx, received) end
+local function toBeNaN(ctx, received)
+	assertType(received, "number")
```

Comment:
Please do not throw any errors based on the received value. Errors should be thrown only by invalid usage of the matchers in testing instructions, not from the code we're testing.